server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -993,24 +993,27 @@ private boolean tryRelocateShard(ModelNode minNode, ModelNode maxNode, String id
                 }
                 ShardRouting candidate = null;
                 final AllocationDeciders deciders = allocation.deciders();
+                final List<ShardRouting> shardRoutings = new ArrayList<>(index.numShards());
                 for (ShardRouting shard : index) {
```

Comment:
I think this could now translate into a nice stream one liner?

```
index.stream().filter(ShardRouting::started).filter(maxNode::containsShard).sorted(Comparator.comparing(ShardRouting::id).reversed())
```

then either collect or do the final filtering and findFirst?

Reply:
We can't use `findFirst` since we also need to capture the decision from the first matching shard; however nor do we need to collect into a list: 8971daab704814f21f931250a0f188c048544675.

(Of course `.sorted()` has to allocate an array anyway, but I think that `Collectors#toList` always copies its input even if it's already an array, so at least there's that)